### PR TITLE
[usm] Refactor telemetry code

### PR DESCRIPTION
**Pull request by p-lambert**
### What does this PR do?

* Simplify API
* Treat all counters as monotonic
* Add reporter for prometheus metrics
* Remove expvar support
* Add helper for building log entries
* Improve debug endpoint

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/encoding/usm.go [hunk 1]
```diff
@@ -168,8 +168,6 @@ func (bc *USMConnectionIndex[K, V]) Close() {
 
 		telemetry.NewMetric(
 			fmt.Sprintf("usm.%s.orphan_aggregations", bc.protocol),
-			telemetry.OptMonotonic,
```
Comment (guyarb) [inline pkg/network/encoding/usm.go:171]: why removing monotonic?

File: pkg/network/usm/monitor.go [hunk 2]
```diff
@@ -321,6 +322,7 @@ func (m *Monitor) GetProtocolStats() map[protocols.ProtocolType]interface{} {
 		// Update update time
 		now := time.Now().Unix()
 		m.lastUpdateTime.Swap(now)
+		telemetry.ReportPrometheus()
```
Comment (guyarb) [inline pkg/network/usm/monitor.go:325]: do we need to actively call that? shouldn't it be pulled from an external component?
What will happen if there is no prometheus?

File: pkg/network/protocols/http/telemetry.go [hunk 3]
```diff
@@ -8,15 +8,13 @@
...
-	"time"
-
 	libtelemetry "github.com/DataDog/datadog-agent/pkg/network/protocols/telemetry"
 	"github.com/DataDog/datadog-agent/pkg/util/log"
-	"go.uber.org/atomic"
 )
 
 type Telemetry struct {
-	LastCheck                                   *atomic.Int64
+	metricGroup *libtelemetry.MetricGroup
```
Comment (guyarb) [inline pkg/network/protocols/http/telemetry.go:16]: can you document? metricGroup does not explain its purpose

File: pkg/network/protocols/http/telemetry.go [hunk 4]
```diff
@@ -27,31 +25,24 @@ type Telemetry struct {
...
-	t := &Telemetry{
-		LastCheck:    atomic.NewInt64(time.Now().Unix()),
-		hits1XX:      metricGroup.NewMetric("hits1xx"),
-		hits2XX:      metricGroup.NewMetric("hits2xx"),
-		hits3XX:      metricGroup.NewMetric("hits3xx"),
-		hits4XX:      metricGroup.NewMetric("hits4xx"),
-		hits5XX:      metricGroup.NewMetric("hits5xx"),
-		aggregations: metricGroup.NewMetric("aggregations"),
+	return &Telemetry{
+		metricGroup: metricGroup,
```
Comment (guyarb) [inline pkg/network/protocols/http/telemetry.go:31]: wondering what's the use to have both the metric group and the actual metrics inside the telemetry object

File: pkg/network/protocols/kafka/telemetry.go [hunk 5]
```diff
@@ -27,37 +23,19 @@ func NewTelemetry() *Telemetry {
 	metricGroup := libtelemetry.NewMetricGroup(
 		"usm.kafka",
 		libtelemetry.OptExpvar,
-		libtelemetry.OptMonotonic,
 	)
 
-	t := &Telemetry{
-		then: atomic.NewInt64(time.Now().Unix()),
-
+	return &Telemetry{
```
Comment (guyarb) [inline pkg/network/protocols/kafka/telemetry.go:28]: you didn't set metricGroup here, it will cause a crash in `log`

File: pkg/network/protocols/telemetry/delta.go [hunk 6]
```diff
@@ -66,14 +64,6 @@ func (c *clientState) ValueFor(m *Metric) int64 {
...
 
 	prev := c.prevValues[name]
-	if prev > current {
-		// let the library client know if this metric is being misconfigured
-		log.Debugf(
-			"error: metric %q is not growing monotonically but it was instantiated with `OptMonotonic`",
-			name,
-		)
-	}
-
```
Comment (guyarb) [inline pkg/network/protocols/telemetry/delta.go:76]: why did you remove the check?

File: pkg/network/protocols/telemetry/metric.go [hunk 7]
```diff
@@ -9,34 +9,37 @@ import (
 	"encoding/json"
 	"strings"
 
-	"github.com/DataDog/datadog-agent/pkg/util/log"
 	"go.uber.org/atomic"
+	"k8s.io/apimachinery/pkg/util/sets"
```
Comment (guyarb) [inline pkg/network/protocols/telemetry/metric.go:13]: we have StringsSet in the project, unless there is a functionality required that is missing, I think we should stick with less type duplications

File: pkg/network/protocols/telemetry/metric.go [hunk 8]
```diff
@@ -104,22 +97,25 @@ func (m *Metric) MarshalJSON() ([]byte, error) {
...
-	// Tags are always sorted
-	for i := range m.tags {
-		if m.tags[i] != other.tags[i] {
-			return false
-		}
+func parseMetricType(opts sets.String) metricType {
+	defer func() {
+		// remove type parameters from the options set
+		opts.Delete(OptGauge)
+		opts.Delete(OptCounter)
```
Comment (guyarb) [inline pkg/network/protocols/telemetry/metric.go:113]: unsure why is it here

**Review by guyarb — CHANGES_REQUESTED** [review]

File: pkg/network/protocols/telemetry/metric.go (hunk 7 above)
Comment (p-lambert) [reply pkg/network/protocols/telemetry/metric.go:13]: I'm not adding a new dependency to the project and our own implementation of StringSet doesn't have everything I need such as `IsSuperSet`

File: pkg/network/encoding/usm.go (hunk 1 above)
Comment (p-lambert) [reply pkg/network/encoding/usm.go:171]: I'm only simplifying things. Counters are now always monotonic. We never had the need for a non-monotonic counter, so I'm removing some cruft related to this. I plan to explicitly mention in the documentation once I make the PR officially ready for review

File: pkg/network/protocols/http/telemetry.go (hunk 3 above)
Comment (p-lambert) [reply pkg/network/protocols/http/telemetry.go:16]: There is documentation here explaining what a `MetricGroup` does:
https://github.com/DataDog/datadog-agent/blob/pedro/telemetry-changes/pkg/network/protocols/telemetry/metric_group.go
Please let me know if that clarifies things.

File: pkg/network/protocols/http/telemetry.go (hunk 4 above)
Comment (p-lambert) [reply pkg/network/protocols/http/telemetry.go:31]: The main motivation here is to use `MetricGroup.Summary()` which simplifies the logging code below a great deal.

File: pkg/network/protocols/telemetry/delta.go (hunk 6 above)
Comment (p-lambert) [reply pkg/network/protocols/telemetry/delta.go:76]: Because counters now are always monotonic

File: pkg/network/protocols/telemetry/metric.go [hunk 9]
```diff
@@ -57,16 +60,24 @@ func NewMetric(name string, tagsAndOptions ...string) *Metric {
...
+	}
+
 	m.value.Store(v)
 }
 
 // Add value atomically
 func (m *Metric) Add(v int64) {
+	if v < 0 {
+		return
+	}
```
[the remark this answers is not in the document]
Comment (p-lambert) [reply pkg/network/protocols/telemetry/metric.go:79]: Agreed, I'll improve this

File: pkg/network/protocols/kafka/telemetry.go (hunk 5 above)
Comment (p-lambert) [reply pkg/network/protocols/kafka/telemetry.go:28]: Good catch! 🙇

File: pkg/network/protocols/telemetry/metric.go (hunk 8 above)
Comment (p-lambert) [reply pkg/network/protocols/telemetry/metric.go:113]: Yeah, I'm going to remove these type arguments and have a `Counter` and `Gauge` type instead

**Review by hmahmood — APPROVED** [review]

File: cmd/system-probe/modules/network_tracer.go [hunk 10]
```diff
@@ -238,10 +238,7 @@ func (nt *networkTracer) Register(httpMux *module.Router) error {
 		utils.WriteAsJSON(w, cache)
 	})
 
-	httpMux.HandleFunc("/debug/telemetry", func(w http.ResponseWriter, req *http.Request) {
-		metrics := telemetry.GetMetrics()
-		utils.WriteAsJSON(w, metrics)
-	})
+	httpMux.HandleFunc("/debug/usm_telemetry", telemetry.Handler)
```
Comment (guyarb) [inline cmd/system-probe/modules/network_tracer.go:241]: why did you change the endpoint? what will it break?

File: pkg/network/usm/ebpf_gotls.go [hunk 11]
```diff
@@ -423,7 +423,7 @@ func (p *GoTLSProgram) hookNewBinary(binID binaryID, binPath string, pid pid, bi
 
 	elapsed := time.Since(start)
 
-	p.binAnalysisMetric.Set(elapsed.Milliseconds())
+	p.binAnalysisMetric.Add(elapsed.Milliseconds())
```
Comment (guyarb) [inline pkg/network/usm/ebpf_gotls.go:426]: Can you show an example after that change?

File: pkg/network/protocols/http/telemetry.go (hunk 4 above)
Comment (guyarb) [reply pkg/network/protocols/http/telemetry.go:31]: I think that is what I missed [here](https://github.com/DataDog/datadog-agent/pull/18151/files#r1261642215)

Comment (guyarb) [reply pkg/network/protocols/http/telemetry.go:31]: Let's add that for the struct

File: pkg/network/protocols/http/telemetry.go [hunk 12]
```diff
@@ -8,50 +8,41 @@
...
+
+		hits1XX:      metricGroup.NewCounter("hits", "status:100", libtelemetry.OptPrometheus),
+		hits2XX:      metricGroup.NewCounter("hits", "status:200", libtelemetry.OptPrometheus),
+		hits3XX:      metricGroup.NewCounter("hits", "status:300", libtelemetry.OptPrometheus),
+		hits4XX:      metricGroup.NewCounter("hits", "status:400", libtelemetry.OptPrometheus),
+		hits5XX:      metricGroup.NewCounter("hits", "status:500", libtelemetry.OptPrometheus),
```
Comment (guyarb) [inline pkg/network/protocols/http/telemetry.go:37]: should we put `status:2xx` instead? will it break anything internally?

**Review by guyarb — APPROVED** [review]

**Review by guyarb — APPROVED** [review]
